core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/DataSource.scala

```diff
@@ -30,26 +34,46 @@ class SlickDataSource(val databaseConfig: DatabaseConfig[JdbcProfile])(implicit
     database.run(f(dataAccess).transactionally.withTransactionIsolation(isolationLevel))
   }
 
-  def createTempTable = {
+  def createEntityAttributeTempTable = {
     sql"""call createEntityAttributeTempTable()""".as[Boolean]
   }
 
-  def dropTempTable = {
+  def dropEntityAttributeTempTable = {
     sql"""call dropEntityAttributeTempTable""".as[Boolean]
   }
 
-  // creates the ENTITY_ATTRIBUTE_TEMP for use by this transaction, executes the transaction, drops the temp table
-  def inTransactionWithAttrTempTable[T](f: (DataAccess) => ReadWriteAction[T], isolationLevel: TransactionIsolation = TransactionIsolation.RepeatableRead): Future[T] = {
+  def createWorkspaceAttributeTempTable = {
+    sql"""call createWorkspaceAttributeTempTable()""".as[Boolean]
+  }
+
+  def dropWorkspaceAttributeTempTable = {
+    sql"""call dropWorkspaceAttributeTempTable""".as[Boolean]
+  }
+
+  // creates the ENTITY_ATTRIBUTE_TEMP, WORSKPACE_ATTRIBUTE_TEMP, or both as specified by tempTableTypes  for use by this transaction, executes the transaction
+  def inTransactionWithAttrTempTable[T](f: (DataAccess) => ReadWriteAction[T], tempTableTypes: Set[AttributeTempTableType.Value], isolationLevel: TransactionIsolation = TransactionIsolation.RepeatableRead): Future[T] = {
 
     val callerAction = f(dataAccess).transactionally.withTransactionIsolation(isolationLevel)
 
-    val callerActionWithTempTables = (for {
-      _ <- dropTempTable
-      _ <- createTempTable
-      origResult <- callerAction
+    val entityTemp = tempTableTypes.contains(AttributeTempTableType.Entity)
+    val workspaceTemp = tempTableTypes.contains(AttributeTempTableType.Workspace)
+
+    val callerActionWithTempTables = for {
+      _ <- if (entityTemp) dropEntityAttributeTempTable else DBIO.successful()
+      _ <- if (workspaceTemp) dropWorkspaceAttributeTempTable else DBIO.successful()
+      _ <- if (entityTemp) createEntityAttributeTempTable else DBIO.successful()
+      _ <- if (workspaceTemp) createWorkspaceAttributeTempTable else DBIO.successful()
+      origResult <- callerAction.asTry
+      _ <- if (entityTemp) dropEntityAttributeTempTable else DBIO.successful()
+      _ <- if (workspaceTemp) dropWorkspaceAttributeTempTable else DBIO.successful()
     } yield {
-      origResult
-    }).andFinally(dropTempTable)
+      origResult match {
+        case scala.util.Success(s) => s
+        case scala.util.Failure(ex) => throw ex
+      }
+    }
+
+
```

Comment:
nit: lots of whitespace here

also, I suggest changing the log message at line 80 to something like:
```
logger.error(s"Transaction with temporary tables failed for (${tempTableTypes.mkString(",")}). Message: ${t.getMessage}")
```